### PR TITLE
Uncertainty: Conformal Prediction V2 - add .conformal_predict() and remove .conformalize() in forecaster.py

### PR DESCRIPTION
**Pull request by Kevin-Chen0**
Resolves a task in Issue #1017.

In forecaster.py, add `.conformal_predict()` method that combines the `.conformalize()` and `.predict()`, and will need to input most of the calibration and test sets for split CP. With this implemented, `.conformalize()` will then be removed. Also, the `config_conformal` will no longer be necessary and will be removed.

For future bootstrapped (e.g. Jackknife+ and CV+ based) CP methods, a calibration set will not be needed, which is why calibration_df is set to None as default in conformal_predict().

### CODE REVIEW

File: neuralprophet/forecaster.py [hunk 1]
```diff
@@ -3021,15 +3006,17 @@ def _reshape_raw_predictions_to_forecst_df(self, df, predicted, components):
                         df_forecast = pd.concat([df_forecast, yhat_df], axis=1, ignore_index=False)
         return df_forecast
 
-    def conformalize(self, df_cal, alpha, method="naive", plotting_backend="default"):
-        """Apply a given conformal prediction technique to get the uncertainty prediction intervals (or q-hats).
+    def conformal_predict(self, df, alpha, calibration_df=None, method="naive", plotting_backend="default"):
```
Comment (noxan) [inline neuralprophet/forecaster.py:3009]: @karl-richter I believe you wanted to add a few more parameters to the regular `predict` function - not sure how we deal with multiple variants of the prediction function, we could replicate the additional parameters and pass them through, thoughts?

File: neuralprophet/forecaster.py [hunk 2]
```diff
@@ -3045,21 +3032,22 @@ def conformalize(self, df_cal, alpha, method="naive", plotting_backend="default"
...
-                df_cal, alpha, method, self.config_train.quantiles, plotting_backend
-            )
-
-    # def conformalize_predict(self, df, df_cal, alpha, method="naive"):
-    #     self.conformalize(df_cal, alpha, method)
-    #     df_forecast = self.predict(df)
-    #     return df_forecast
+        q_hats = conformalize(df_cal, alpha, method, self.config_train.quantiles, plotting_backend)
+        # predict
+        df = self.predict(df)
```
Comment (noxan) [inline neuralprophet/forecaster.py:3041]: we should allow users to pass arguments to the prediction function, also see comment above.

**Comment by noxan** [comment]
Not really sure which is the best API, having an additional setup method or wrapping the predict method - will we have more comparable cases in the future (like other uncertainty prediction modes)? would be great if we can find the most generic approach.

I like the reduction of code 👍 

Added a few minor remarks. Happy to hear your thoughts @Kevin-Chen0, @ourownstory and @karl-richter.

File: neuralprophet/forecaster.py (hunk 2 above)
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py:3041]: How about pass in **kwargs into `conformal_predict()`, which will then be passed in the `self.predict()`?

File: neuralprophet/forecaster.py [hunk 3]
```diff
@@ -3045,21 +3032,22 @@ def conformalize(self, df_cal, alpha, method="naive", plotting_backend="default"
...
+        df["qhat1"] = q_hats[0]
+        if method == "naive":
+            df["yhat1 - qhat1"] = df["yhat1"] - q_hats[0]
+            df["yhat1 + qhat1"] = df["yhat1"] + q_hats[0]
+        else:  # method == "cqr"
```
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py:3046]: You mean throw `ValueError`? If that is the conversion in forecaster.py then sure.

File: neuralprophet/forecaster.py (hunk 1 above)
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py:3009]: Pass in **kwargs?

**Comment by Kevin-Chen0** [comment]
> Not really sure which is the best API, having an additional setup method or wrapping the predict method - will we have more comparable cases in the future (like other uncertainty prediction modes)? would be great if we can find the most generic approach.
> 
> I like the reduction of code +1
> 
> Added a few minor remarks. Happy to hear your thoughts @Kevin-Chen0, @ourownstory and @karl-richter.

I prefer this PR of wrapping the predict method as `conformal_predict()`, because all of my conformal prediction code sit nicely into `conformal_predict()` w/o intruding into other codes. If I have multiple setup methods like `conformalize()` (as the current setup), then my conformal code will be spread out in forecaster.py, including parts in `predict()`, in `config_conformal`, etc.

That is how with this PR, I can achieve a reduction in the code. Nevertheless, I'm happy to hear other people's thoughts.

File: neuralprophet/forecaster.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py]: Implemented.

File: neuralprophet/forecaster.py (hunk 1 above)
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py]: Implemented.

File: neuralprophet/forecaster.py (hunk 2 above)
Comment (Kevin-Chen0) [reply neuralprophet/forecaster.py]: Implemented.

File: neuralprophet/forecaster.py [hunk 4]
```diff
@@ -3021,13 +3006,17 @@ def _reshape_raw_predictions_to_forecst_df(self, df, predicted, components):
                         df_forecast = pd.concat([df_forecast, yhat_df], axis=1, ignore_index=False)
         return df_forecast
 
-    def conformalize(self, df_cal, alpha, method="naive", plotting_backend="default"):
-        """Apply a given conformal prediction technique to get the uncertainty prediction intervals (or q-hats).
+    def conformal_predict(
+        self, df, calibration_df=None, alpha=0.1, method="naive", plotting_backend="default", **kwargs
```
Comment (noxan) [inline neuralprophet/forecaster.py:3010]: Not sure myself: Should we better explicitly define the parameters here? If people use the docstrings and code reference to understand this method they won't get far like this. 🤔

File: neuralprophet/forecaster.py [hunk 5]
```diff
@@ -3044,22 +3033,27 @@ def conformalize(self, df_cal, alpha, method="naive", plotting_backend="default"
                     * ``plotly``: Use the plotly backend for plotting
                     * ``matplotlib``: Use matplotlib backend for plotting
                     * ``default`` (default): Use matplotlib backend for plotting
+            kwargs : dict
+                additional predict parameters for test df
         """
-        df_cal = self.predict(df_cal)
+        # conformalize
+        df_cal = self.predict(calibration_df)
```
Comment (noxan) [inline neuralprophet/forecaster.py:3040]: What does `df_cal` stand for? Could we find a more verbose and clear name instead?

**Review by noxan — APPROVED** [review]
LGTM, would be curious to hear your thoughts on the two minor remarks @Kevin-Chen0 

In the long term I am not too happy with having two separate `predict` functions and so different APIs for uncertainty methods - let's discuss this somewhen later.

**Pull request merged**